src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -324,7 +322,7 @@ protected void spillStructBox(Binding.Builder bindings, MemoryLayout layout) {
                         for (int i = 0; i < group.memberLayouts().size(); i++) {
                             VMStorage storage = regs[i];
                             final long size = group.memberLayouts().get(i).byteSize();
-                            Class<?> type = SharedUtils.primitiveCarrierForSize(size);
+                            Class<?> type = SharedUtils.primitiveCarrierForSize(size, false);
```

Comment:
Not sure this is ok - this is a struct of floats passed by register?

Reply:
Yes, I found that that the native code passes float-only structs in float registers, but we were trying to load it using an int carrier, which maps to an int register in the Java calling convention. Besides being inefficient to have to move between these registers, it is not possible to move between float and int registers directly in x86, which would complicate the shuffling process by having to go through memory.

By using float/double as a carrier here instead, the upcall handle expects it's argument to be in a float register as well, and we don't have to move it between registers.

Reply:
Ah, wait this is AArch64. I've kept this as is so far (it was previously using ints as a carrier as well). You are probably right that this is incorrect.

Currently it doesn't matter though since we always go through an intermediate buffer (AArch64 doesn't support specialized upcalls yet). I simply didn't attempt to pick the correct type here yet.

Reply:
Yes - this is Aarch, the STRUCT_HFA case which AFAIK should be structs passed by FP registers.